File: app/Sources/Tideline/Rule.swift
import Foundation

/// Rules that claim a file by what it is called, or by where it came from.
///
/// A type rule asks what a file *is*, which settles a `.dmg` and settles nothing
/// about an invoice: an invoice is a PDF like every other PDF. What marks one out
/// is its name, or the URL it arrived from — so a rule is a folder and a list of
/// tests, and the rule says whether one of them matching is enough or all of
/// them have to agree.
///
/// Rules are consulted before the type rules and after everything that decides
/// whether a file moves at all. They answer *where*, never *when*.
///
/// The contract is `docs/behaviour.md`; the cases are `fixtures/sweep-cases.json`.

/// What a test looks at.
enum RuleField: String, Codable, CaseIterable, Identifiable {
    /// The entry's own name, extension included.
    case name
    /// Each URL the download was recorded as arriving from, in turn.
    case whereFrom = "where_from"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .name: return "Name"
        case .whereFrom: return "Downloaded from"
        }
    }

    /// How a match reads back in a list — `name contains invoice`, `download
    /// URL contains stripe.com`. Lower case, because it is a fragment of a
    /// sentence rather than a heading, and a noun rather than the picker's
    /// "Downloaded from", which only reads as a label with a value beside it.
    var shortLabel: String {
        switch self {
        case .name: return "name"
        case .whereFrom: return "download URL"
        }
    }

    /// A pattern in this field, for the one operator that takes a whole glob.
    var placeholder: String {
        switch self {
        case .name: return "*invoice*"
        case .whereFrom: return "*stripe*"
        }
    }

    /// The words rather than the pattern, for the other four.
    var valueHint: String {
        switch self {
        case .name: return "invoice"
        case .whereFrom: return "stripe.com"
        }
    }
}

/// How a pattern is meant, spelled out.
///
/// The stored grammar is still a glob — `*invoice*` — because that is what the
/// contract says and what the Rust engine reads. This is the same thing said in
/// words, so the editor can offer *contains* rather than asking someone to know
/// where the stars go. It is presentation, never storage: a test round-trips
/// through `RuleTest.pattern` and nothing else.
enum RuleOperator: String, CaseIterable, Identifiable {
    /// `*value*`
    case contains
    /// `value`
    case exact
    /// `value*`
    case startsWith
    /// `*value`
    case endsWith
    /// The glob itself, for anything the four above cannot say.
    case glob

    var id: String { rawValue }

    var label: String {
        switch self {
        case .contains: return "contains"
        case .exact: return "is"
        case .startsWith: return "starts with"
        case .endsWith: return "ends with"
        case .glob: return "matches pattern"
        }
    }

    /// The pattern this operator and value add up to.
    func pattern(for value: String) -> String {
        let value = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty else { return "" }
        switch self {
        case .contains: return "*\(value)*"
        case .exact: return value
        case .startsWith: return "\(value)*"
        case .endsWith: return "*\(value)"
        case .glob: return value
        }
    }

    /// The same condition said a different way, when the operator changes.
    ///
    /// Switching *to* a pattern hands over the glob the words added up to, so
    /// there is something to edit rather than a field that quietly lost its
    /// stars; switching *away* from one reads the words back out, so `*invoice*`
    /// under *contains* does not become `**invoice**`.
    static func reword(_ value: String, from old: RuleOperator, to new: RuleOperator) -> String {
        guard old != new else { return value }
        if new == .glob { return old.pattern(for: value) }
        if old == .glob { return read(value).1 }
        return value
    }

    /// A stored pattern read back as an operator and the words inside it.
    ///
    /// A pattern whose *middle* holds a wildcard cannot be said in words, so it
    /// reads back as itself under *matches pattern* rather than being described
    /// as something it is not. That is also what someone who typed a star into
    /// a *contains* field gets the next time they open the rule — the glob they
    /// actually saved, not a tidied-up version of it.
    static func read(_ pattern: String) -> (RuleOperator, String) {
        let pattern = pattern.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !pattern.isEmpty else { return (.contains, "") }

        let leading = pattern.hasPrefix("*")
        let trailing = pattern.hasSuffix("*") && pattern.count > 1
        var core = pattern
        if leading { core.removeFirst() }
        if trailing { core.removeLast() }

        guard !core.isEmpty, !core.contains(where: { "*?[".contains($0) }) else {
            return (.glob, pattern)
        }

        switch (leading, trailing) {
        case (true, true): return (.contains, core)
        case (true, false): return (.endsWith, core)
        case (false, true): return (.startsWith, core)
        case (false, false): return (.exact, core)
        }
    }
}

/// One thing a rule asks about a file.
struct RuleTest: Codable, Identifiable, Equatable {
    var id = UUID()
    var field: RuleField = .name
    /// The skip list's grammar — `*`, `?`, `[a-z]`, `[!abc]` — so `*invoice*`
    /// means "contains" without needing a second operator.
    var pattern: String = ""
    /// Off by default, unlike the skip list. A skip pattern names a file you can
    /// already see in the folder; a rule is a hunt for a word that could have
    /// been typed either way, and a rule that quietly stops matching is not
    /// something anyone notices.
    var matchCase: Bool = false

    init(id: UUID = UUID(), field: RuleField = .name, pattern: String = "", matchCase: Bool = false) {
        self.id = id
        self.field = field
        self.pattern = pattern
        self.matchCase = matchCase
    }
}

extension RuleTest {
    /// The stored pattern said in words, for the editor's two popups.
    var spelledOut: (op: RuleOperator, value: String) { RuleOperator.read(pattern) }

    /// How this test reads in the row above the editor — `name contains
    /// invoice`, `downloaded from stripe.com`. A fragment of a sentence, so it
    /// is lower case and carries no quotes.
    var sentence: String? {
        let (op, value) = spelledOut
        guard !value.isEmpty else { return nil }
        return "\(field.shortLabel) \(op.label) \(value)"
    }

    private enum CodingKeys: String, CodingKey {
        case id, field, pattern, matchCase
    }

    /// Decoded leniently, like everything else the app persists: a test written
    /// by the Windows app carries no `id`, and an unknown field falls back to the
    /// name rather than failing the whole rule list.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(UUID.self, forKey: .id) ?? UUID()
        field = try container.decodeIfPresent(RuleField.self, forKey: .field) ?? .name
        pattern = try container.decodeIfPresent(String.self, forKey: .pattern) ?? ""
        matchCase = try container.decodeIfPresent(Bool.self, forKey: .matchCase) ?? false
    }

    /// Whether this test claims the file.
    ///
    /// An empty pattern claims nothing: a half-written rule in the settings
    /// window would otherwise match everything the moment the field was cleared.
    func matches(_ subject: RuleSubject) -> Bool {
        let pattern = self.pattern.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !pattern.isEmpty else { return false }

        switch field {
        case .name:
            return RuleTest.glob(pattern, matches: subject.name, matchCase: matchCase)
        case .whereFrom:
            // The interesting word is often escaped inside a query string, so the
            // URL is decoded before it is matched: someone who typed
            // `*/facturen/*` means to find `?path=%2Ffacturen%2F`.
            return subject.whereFrom.contains { url in
                RuleTest.glob(pattern, matches: RuleTest.percentDecoded(url), matchCase: matchCase)
            }
        }
    }

    /// `fnmatch`, with the case rule spelled out rather than left to the platform.
    ///
    /// Folding is done here rather than with `FNM_CASEFOLD` because that calls
    /// `tolower` and so answers to the process locale. The contract says ASCII
    /// `A`–`Z` and nothing else, in every locale and on both platforms, so that a
    /// rule cannot mean one thing here and another in the Rust engine.
    static func glob(_ pattern: String, matches name: String, matchCase: Bool) -> Bool {
        guard !matchCase else { return fnmatch(pattern, name, 0) == 0 }
        return fnmatch(asciiFolded(pattern), asciiFolded(name), 0) == 0
    }

    /// `A`–`Z` down to `a`–`z`, and every other scalar left exactly as it is.
    static func asciiFolded(_ text: String) -> String {
        guard text.unicodeScalars.contains(where: { $0.value >= 65 && $0.value <= 90 }) else {
            return text
        }
        var folded = String.UnicodeScalarView()
        for scalar in text.unicodeScalars {
            if scalar.value >= 65, scalar.value <= 90 {
                folded.append(UnicodeScalar(scalar.value + 32) ?? scalar)
            } else {
                folded.append(scalar)
            }
        }
        return String(folded)
    }

    /// `%2F` back to `/`, so a pattern can be written the way the path reads.
    ///
    /// Written out rather than handed to `removingPercentEncoding`, which gives
    /// up on the whole string when one escape is malformed. Anything that is not
    /// a valid escape is left exactly as it was: a lone `%` in a
    /// filename-turned-URL is likelier than a typo in an escape, and mangling it
    /// would lose a character the pattern might be looking for.
    static func percentDecoded(_ text: String) -> String {
        guard text.contains("%") else { return text }

        let bytes = Array(text.utf8)
        var out: [UInt8] = []
        out.reserveCapacity(bytes.count)
        var index = 0

        while index < bytes.count {
            if bytes[index] == UInt8(ascii: "%"), index + 2 < bytes.count,
               let high = hexValue(bytes[index + 1]), let low = hexValue(bytes[index + 2]) {
                out.append(high * 16 + low)
                index += 3
                continue
            }
            out.append(bytes[index])
            index += 1
        }

        // A percent escape can name a byte that is not valid UTF-8 on its own.
        // The pattern could never have matched it anyway, so a replacement
        // character is a better answer than throwing the whole URL away.
        return String(decoding: out, as: UTF8.self)
    }

    private static func hexValue(_ byte: UInt8) -> UInt8? {
        switch byte {
        case UInt8(ascii: "0")...UInt8(ascii: "9"): return byte - UInt8(ascii: "0")
        case UInt8(ascii: "a")...UInt8(ascii: "f"): return byte - UInt8(ascii: "a") + 10
        case UInt8(ascii: "A")...UInt8(ascii: "F"): return byte - UInt8(ascii: "A") + 10
        default: return nil
        }
    }
}

/// Whether a rule needs one of its tests or every one of them.
enum RuleMatch: String, Codable, CaseIterable, Identifiable {
    /// Any one test is enough. What a rule has always meant, and the default.
    case any
    /// Every test has to agree. "A PDF from Stripe" in one rule instead of two.
    case all

    var id: String { rawValue }

    /// Read inside the sentence the editor builds: "Catch a file when *any* of
    /// these is true".
    var label: String { rawValue }
}

/// A folder in the root, and the tests that send files to it.
struct Rule: Codable, Identifiable, Equatable {
    /// Stable across renames, so a rule the user renamed is still the same rule
    /// the next time the app starts.
    var id: String
    /// The destination: a folder name in the root. Stored as `name` because
    /// that is the key both engines have always read — `folder` below is the
    /// same thing under the word the window uses for it.
    var name: String
    /// What the list calls this rule, when that is not simply its folder.
    ///
    /// Two rules can send files to one folder — receipts from Stripe and
    /// receipts from the bank — and a list that shows both as `Receipts` cannot
    /// be reordered with any confidence. Optional, and absent on every rule
    /// written before it existed, so the folder goes on standing in for it.
    var title: String?
    var isEnabled: Bool
    /// Whether one test is enough or all of them have to agree.
    var match: RuleMatch
    var tests: [RuleTest]

    init(
        id: String = UUID().uuidString,
        name: String,
        title: String? = nil,
        isEnabled: Bool = true,
        match: RuleMatch = .any,
        tests: [RuleTest] = []
    ) {
        self.id = id
        self.name = name
        self.title = title
        self.isEnabled = isEnabled
        self.match = match
        self.tests = tests
    }

    /// The destination folder, under the word the window uses for it.
    var folder: String { name }

    /// What to call this rule in a list. A rule that was never given a name of
    /// its own is known by where it sends things, and one that has neither yet
    /// is the rule somebody added a second ago.
    var displayName: String {
        let title = (title ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if !title.isEmpty { return title }
        let folder = name.trimmingCharacters(in: .whitespacesAndNewlines)
        return folder.isEmpty ? "Untitled rule" : folder
    }

    /// Tests with something actually typed in them. An empty one is not a test
    /// yet — it is a row someone has only just added — and under `all` it would
    /// otherwise stop the rule matching anything at all.
    var filledTests: [RuleTest] {
        tests.filter { !$0.pattern.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
    }
}

extension Rule {
    private enum CodingKeys: String, CodingKey {
        case id, name, title, isEnabled, match, tests
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
        name = try container.decodeIfPresent(String.self, forKey: .name) ?? ""
        title = try container.decodeIfPresent(String.self, forKey: .title)
        isEnabled = try container.decodeIfPresent(Bool.self, forKey: .isEnabled) ?? false
        // A rule saved before the setting existed meant `any`, which is also
        // what an unreadable value has to fall back to: `all` narrows what a
        // rule catches, and a rule that quietly stops firing after an upgrade
        // is not something anybody notices.
        match = try container.decodeIfPresent(RuleMatch.self, forKey: .match) ?? .any
        tests = try container.decodeIfPresent([RuleTest].self, forKey: .tests) ?? []
    }

    /// Whether this rule claims the file.
    ///
    /// A rule with no filled-in test claims nothing, which is what a rule
    /// someone has only just added looks like.
    func matches(_ subject: RuleSubject) -> Bool {
        let live = filledTests
        guard !live.isEmpty else { return false }
        switch match {
        case .any: return live.contains { $0.matches(subject) }
        case .all: return live.allSatisfy { $0.matches(subject) }
        }
    }

    /// The test that claimed the file, for a list that has to say why.
    ///
    /// Under `all` every test agreed, so the first one is as good an answer as
    /// any; under `any` it is the one that actually did the claiming.
    func claimingTest(_ subject: RuleSubject) -> RuleTest? {
        guard matches(subject) else { return nil }
        switch match {
        case .any: return filledTests.first { $0.matches(subject) }
        case .all: return filledTests.first
        }
    }

    /// A rule's folder is held to the same standard as a type folder's, since
    /// both end up as a folder in the root.
    static func isValidFolderName(_ name: String) -> Bool {
        TypeRule.isValidFolderName(name)
    }
}

/// What a rule is allowed to ask about one file.
///
/// The where-froms are read the first time a test asks for them and not before:
/// they live in an extended attribute, and a sweep whose rules only look at names
/// should not be reading one per file. A class rather than a struct so the answer
/// is remembered across the tests of every rule in the list.
final class RuleSubject {
    let name: String
    private let url: URL
    private var loaded: [String]?

    init(name: String, url: URL) {
        self.name = name
        self.url = url
    }

    var whereFrom: [String] {
        if let loaded { return loaded }
        let urls = WhereFrom.urls(for: url)
        loaded = urls
        return urls
    }
}

/// Decides which rule, if any, claims a given file, and knows the names of the
/// folders those rules own so filing never files one of them away.
///
/// Built once per sweep from an immutable snapshot of the rules.
struct RuleRouter {
    /// Enabled rules whose folder name is usable, in the order they are tried.
    private let rules: [Rule]
    private let ownedNames: Set<String>

    /// Only enabled rules are given to this. They are tried in order and the
    /// first match wins, which is the order the settings window shows.
    init(rules: [Rule]) {
        var kept: [Rule] = []
        var names: Set<String> = []

        for rule in rules {
            let name = rule.name.trimmingCharacters(in: .whitespacesAndNewlines)
            guard Rule.isValidFolderName(name) else { continue }
            names.insert(name.lowercased())
            var trimmed = rule
            trimmed.name = name
            kept.append(trimmed)
        }

        self.rules = kept
        ownedNames = names
    }

    var isEmpty: Bool { rules.isEmpty }

    /// The folder this file belongs in, or nil to let the type rules and then
    /// the date decide.
    func folderName(for subject: RuleSubject) -> String? {
        rules.first { $0.matches(subject) }?.name
    }

    /// A rule's folder is never itself filed into a dated one.
    func owns(_ name: String) -> Bool {
        ownedNames.contains(name.lowercased())
    }
}

/// Where a download came from, as macOS recorded it.
///
/// Safari, Chrome and the rest write the source URL and the referrer into an
/// extended attribute, and the attribute travels with the file — which is what
/// lets a rule recognise something that was renamed by hand, and lets a rule
/// written today match what arrived last month.
///
/// Windows keeps the same information in the `Zone.Identifier` alternate data
/// stream; `docs/behaviour.md` has the pair.
enum WhereFrom {
    static let attribute = "com.apple.metadata:kMDItemWhereFroms"

    /// Every URL recorded for this file, source first. Empty is normal and never
    /// an error: a file saved out of a mail client, dropped in over AirDrop or
    /// fetched with `curl` carries none.
    static func urls(for url: URL) -> [String] {
        guard let data = attributeData(at: url.path) else { return [] }
        guard
            let plist = try? PropertyListSerialization.propertyList(
                from: data, options: [], format: nil
            ),
            let list = plist as? [Any]
        else { return [] }

        return list.compactMap { $0 as? String }.filter { !$0.isEmpty }
    }

    /// The raw attribute, which is a binary plist holding an array of strings.
    private static func attributeData(at path: String) -> Data? {
        path.withCString { cPath -> Data? in
            let length = getxattr(cPath, attribute, nil, 0, 0, 0)
            guard length > 0 else { return nil }

            var buffer = [UInt8](repeating: 0, count: length)
            let read = buffer.withUnsafeMutableBytes { raw in
                getxattr(cPath, attribute, raw.baseAddress, length, 0, 0)
            }
            guard read > 0 else { return nil }

            return Data(buffer.prefix(read))
        }
    }
}

File: windows/src-tauri/crates/tideline-core/src/rule.rs
//! Rules that claim a file by what it is called, or by where it came from.
//!
//! A type rule asks what a file *is*, which settles a `.msi` and settles
//! nothing about an invoice: an invoice is a PDF like every other PDF. What
//! marks one out is its name, or the URL it arrived from — so a rule is a folder
//! and a list of tests, and the rule says whether one of them matching is
//! enough or all of them have to agree.
//!
//! Rules are consulted before the type rules and after everything that decides
//! whether a file moves at all. They answer *where*, never *when*.
//!
//! The contract is `docs/behaviour.md`; the cases are `fixtures/sweep-cases.json`.

use crate::organizer::Entry;
use crate::settings::is_valid_folder_name;
use crate::skip::glob_match_with;
use serde::{Deserialize, Serialize};

/// What a test looks at.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum Field {
    /// The entry's own name, extension included.
    #[default]
    Name,
    /// Each URL the download was recorded as arriving from, in turn.
    WhereFrom,
}

impl Field {
    pub fn label(self) -> &'static str {
        match self {
            Field::Name => "Name",
            Field::WhereFrom => "Downloaded from",
        }
    }
}

/// One thing a rule asks about a file.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RuleTest {
    #[serde(default)]
    pub field: Field,
    /// The skip list's grammar — `*`, `?`, `[a-z]`, `[!abc]` — so `*invoice*`
    /// means "contains" without needing a second operator.
    pub pattern: String,
    /// Off by default, unlike the skip list on macOS. A skip pattern names a
    /// file you can already see; a rule is a hunt for a word that could have
    /// been typed either way, and a rule that quietly stops matching is not
    /// something anyone notices.
    #[serde(rename = "matchCase", default)]
    pub match_case: bool,
}

impl RuleTest {
    pub fn new(field: Field, pattern: &str) -> Self {
        RuleTest {
            field,
            pattern: pattern.to_string(),
            match_case: false,
        }
    }

    /// Whether this test claims the entry.
    ///
    /// An empty pattern claims nothing. A half-written rule in the settings
    /// window would otherwise match everything the moment the field was
    /// cleared.
    pub fn matches(&self, entry: &Entry) -> bool {
        let pattern = self.pattern.trim();
        if pattern.is_empty() {
            return false;
        }

        match self.field {
            Field::Name => glob_match_with(pattern, &entry.name, self.match_case),
            // The interesting word is often escaped inside a query string, so
            // the URL is decoded before it is matched: someone who typed
            // `*/facturen/*` means to find `?path=%2Ffacturen%2F`.
            Field::WhereFrom => entry
                .where_from
                .iter()
                .any(|url| glob_match_with(pattern, &percent_decode(url), self.match_case)),
        }
    }
}

/// Whether a rule needs one of its tests or every one of them.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum Match {
    /// Any one test is enough. What a rule has always meant, and the default a
    /// rule written before the setting existed falls back to.
    #[default]
    Any,
    /// Every test has to agree.
    All,
}

/// A folder in the root, and the tests that send files to it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Rule {
    /// Stable across renames, so a rule the user renamed is still the same rule
    /// next launch.
    pub id: String,
    /// The destination: a folder name in the root.
    pub name: String,
    /// What a list calls this rule, when that is not simply its folder. The
    /// engine never reads it; it is carried so writing the settings back does
    /// not throw away what the other platform's window put there.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(rename = "isEnabled")]
    pub is_enabled: bool,
    /// Whether one test is enough or all of them have to agree.
    #[serde(rename = "match", default)]
    pub match_mode: Match,
    #[serde(default)]
    pub tests: Vec<RuleTest>,
}

impl Rule {
    pub fn new(id: &str, name: &str, tests: Vec<RuleTest>) -> Self {
        Rule {
            id: id.to_string(),
            name: name.to_string(),
            title: None,
            is_enabled: false,
            match_mode: Match::Any,
            tests,
        }
    }

    /// Whether this rule claims the entry.
    ///
    /// A rule with no filled-in test claims nothing, which is what a rule
    /// someone has only just added looks like. An empty pattern is dropped
    /// before the question is asked rather than answered `false` inside it:
    /// under `all` a half-written test would otherwise stop the rule matching
    /// anything at all.
    pub fn matches(&self, entry: &Entry) -> bool {
        let mut live = self
            .tests
            .iter()
            .filter(|test| !test.pattern.trim().is_empty());
        match self.match_mode {
            Match::Any => live.any(|test| test.matches(entry)),
            Match::All => {
                let mut live = live.peekable();
                live.peek().is_some() && live.all(|test| test.matches(entry))
            }
        }
    }
}

/// Decides which rule, if any, claims a given file, and knows the names of the
/// folders those rules own so filing never files one of them away.
///
/// Built once per sweep from an immutable snapshot of the rules.
pub struct RuleRouter {
    /// Enabled rules whose folder name is usable, in the order they are tried.
    rules: Vec<Rule>,
    /// Lower-cased folder names, for the "is this a destination?" question.
    owned_names: Vec<String>,
}

impl RuleRouter {
    /// Only enabled rules should be handed to this. They are tried in order and
    /// the first match wins, which is the order the settings window shows.
    pub fn new(rules: &[Rule]) -> Self {
        let mut kept: Vec<Rule> = Vec::new();
        let mut owned_names: Vec<String> = Vec::new();

        for rule in rules {
            let name = rule.name.trim();
            if !is_valid_folder_name(name) {
                continue;
            }
            let lowered = name.to_lowercase();
            if !owned_names.contains(&lowered) {
                owned_names.push(lowered);
            }
            kept.push(Rule {
                name: name.to_string(),
                ..rule.clone()
            });
        }

        RuleRouter {
            rules: kept,
            owned_names,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.rules.is_empty()
    }

    /// The folder this file belongs in, or `None` to let the type rules and
    /// then the date decide.
    pub fn folder_name(&self, entry: &Entry) -> Option<&str> {
        self.rules
            .iter()
            .find(|rule| rule.matches(entry))
            .map(|rule| rule.name.as_str())
    }

    /// A rule's folder is never itself filed into a dated one.
    pub fn owns(&self, name: &str) -> bool {
        self.owned_names.contains(&name.to_lowercase())
    }
}

/// `%2F` back to `/`, so a pattern can be written the way the path reads.
///
/// Anything that is not a valid escape is left exactly as it was: a lone `%` in
/// a filename-turned-URL is far more likely than a typo in an escape, and
/// mangling it would lose a character the pattern might be looking for.
pub fn percent_decode(text: &str) -> String {
    if !text.contains('%') {
        return text.to_string();
    }

    let bytes = text.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut i = 0;

    while i < bytes.len() {
        if bytes[i] == b'%' && i + 2 < bytes.len() {
            if let (Some(high), Some(low)) = (hex(bytes[i + 1]), hex(bytes[i + 2])) {
                out.push(high * 16 + low);
                i += 3;
                continue;
            }
        }
        out.push(bytes[i]);
        i += 1;
    }

    // A percent escape can name a byte that is not valid UTF-8 on its own. The
    // pattern could never have matched it anyway, so a replacement character is
    // a better answer than throwing the whole URL away.
    String::from_utf8_lossy(&out).into_owned()
}

fn hex(byte: u8) -> Option<u8> {
    match byte {
        b'0'..=b'9' => Some(byte - b'0'),
        b'a'..=b'f' => Some(byte - b'a' + 10),
        b'A'..=b'F' => Some(byte - b'A' + 10),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Local;

    fn entry(name: &str, where_from: &[&str]) -> Entry {
        Entry {
            name: name.into(),
            is_dir: false,
            stamp: Some(Local::now()),
            size: 1024,
            settled: true,
            hidden: false,
            where_from: where_from.iter().map(|s| s.to_string()).collect(),
        }
    }

    fn enabled(id: &str, name: &str, tests: Vec<RuleTest>) -> Rule {
        let mut rule = Rule::new(id, name, tests);
        rule.is_enabled = true;
        rule
    }

    #[test]
    fn a_name_test_ignores_case_by_default() {
        let test = RuleTest::new(Field::Name, "invoice-*");
        assert!(test.matches(&entry("Invoice-0005.pdf", &[])));
        assert!(test.matches(&entry("INVOICE-0005.PDF", &[])));
        assert!(!test.matches(&entry("statement-0005.pdf", &[])));
    }

    #[test]
    fn match_case_compares_exactly() {
        let mut test = RuleTest::new(Field::Name, "Invoice-*");
        test.match_case = true;
        assert!(test.matches(&entry("Invoice-0005.pdf", &[])));
        assert!(!test.matches(&entry("INVOICE-0005.pdf", &[])));
    }

    #[test]
    fn folding_is_ascii_only() {
        // Ä and ä stay different words. Full Unicode folding is
        // locale-dependent, and the two engines have to agree.
        let test = RuleTest::new(Field::Name, "factuur-ä*");
        assert!(test.matches(&entry("Factuur-ä-0042.pdf", &[])));
        assert!(!test.matches(&entry("Factuur-Ä-0042.pdf", &[])));
    }

    #[test]
    fn a_where_from_test_reads_every_recorded_url() {
        let test = RuleTest::new(Field::WhereFrom, "*invoice*");
        // The identifying word is in the path, not the host — matching only the
        // hostname would miss this entirely.
        assert!(test.matches(&entry(
            "logitech-palm-rest.pdf",
            &[
                "https://api.global-e.com/Document/CustomerReceiptInvoice?documentParam=KqND",
                "https://www.logitech.com/",
            ],
        )));
        assert!(!test.matches(&entry("manual.pdf", &["https://www.logitech.com/"])));
    }

    #[test]
    fn a_url_is_decoded_before_it_is_matched() {
        let test = RuleTest::new(Field::WhereFrom, "*/facturen/*");
        assert!(test.matches(&entry(
            "20260702.pdf",
            &["https://portal.example.com/download?path=%2Ffacturen%2Ffactuur-42.pdf"],
        )));
        assert!(!test.matches(&entry(
            "bon.pdf",
            &["https://portal.example.com/download?path=%2Fbonnen%2Fbon-7.pdf"],
        )));
    }

    #[test]
    fn a_file_with_no_recorded_source_matches_nothing_on_that_field() {
        let test = RuleTest::new(Field::WhereFrom, "*invoice*");
        // Not an error, and not a match: the rule falls through to its next
        // test, and the file to the next rule.
        assert!(!test.matches(&entry("Invoice-0005.pdf", &[])));
    }

    #[test]
    fn tests_are_ord() {
        let rule = enabled(
            "invoices",
            "Invoices",
            vec![
                RuleTest::new(Field::Name, "*invoice*"),
                RuleTest::new(Field::WhereFrom, "*stripe*"),
            ],
        );
        assert!(rule.matches(&entry("Invoice-0005.pdf", &[])));
        assert!(rule.matches(&entry(
            "2026-07.pdf",
            &["https://stripe-upload-api.s3.aws/x"]
        )));
        assert!(!rule.matches(&entry("holiday.jpg", &["https://example.com/"])));
    }

    #[test]
    fn a_rule_set_to_all_needs_every_test() {
        let mut rule = enabled(
            "stripe-invoices",
            "Stripe invoices",
            vec![
                RuleTest::new(Field::Name, "*invoice*"),
                RuleTest::new(Field::WhereFrom, "*stripe.com*"),
            ],
        );
        rule.match_mode = Match::All;

        assert!(rule.matches(&entry(
            "invoice-0042.pdf",
            &["https://files.stripe.com/invoices/acct_1/0042"]
        )));
        // Each of these satisfies one test and not the other, which under `any`
        // would have been enough for both of them.
        assert!(!rule.matches(&entry(
            "invoice-from-the-bank.pdf",
            &["https://bank.example.com/"]
        )));
        assert!(!rule.matches(&entry(
            "products.csv",
            &["https://files.stripe.com/exports/products"]
        )));
    }

    #[test]
    fn a_half_written_test_does_not_stop_a_rule_set_to_all() {
        let mut rule = enabled(
            "invoices",
            "Invoices",
            vec![
                RuleTest::new(Field::Name, "*invoice*"),
                RuleTest::new(Field::WhereFrom, "   "),
            ],
        );
        rule.match_mode = Match::All;
        assert!(rule.matches(&entry("invoice-0042.pdf", &[])));
    }

    #[test]
    fn a_rule_set_to_all_with_nothing_written_down_claims_nothing() {
        let mut rule = enabled("new", "Invoices", vec![RuleTest::new(Field::Name, "")]);
        rule.match_mode = Match::All;
        assert!(!rule.matches(&entry("anything.pdf", &[])));
    }

    #[test]
    fn a_rule_stored_without_a_match_means_any() {
        // Every rule written before the setting existed, and every rule the
        // Windows window has ever saved.
        let rule: Rule = serde_json::from_str(
            r#"{"id":"invoices","name":"Invoices","isEnabled":true,
                "tests":[{"field":"name","pattern":"*invoice*"},
                         {"field":"where_from","pattern":"*stripe.com*"}]}"#,
        )
        .expect("a rule with no match still reads");
        assert_eq!(rule.match_mode, Match::Any);
        assert!(rule.matches(&entry("invoice-0042.pdf", &[])));
    }

    #[test]
    fn an_empty_pattern_claims_nothing() {
        let test = RuleTest::new(Field::Name, "   ");
        assert!(!test.matches(&entry("anything.pdf", &[])));
    }

    #[test]
    fn a_rule_with_no_tests_claims_nothing() {
        let rule = enabled("new", "Invoices", vec![]);
        let router = RuleRouter::new(&[rule]);
        assert_eq!(router.folder_name(&entry("Invoice-0005.pdf", &[])), None);
        // It still owns its folder, so the folder is not filed away while the
        // rule is being written.
        assert!(router.owns("Invoices"));
    }

    #[test]
    fn the_first_rule_that_matches_wins() {
        let router = RuleRouter::new(&[
            enabled(
                "receipts",
                "Receipts",
                vec![RuleTest::new(Field::Name, "*receipt*")],
            ),
            enabled(
                "invoices",
                "Invoices",
                vec![RuleTest::new(Field::Name, "*invoice*")],
            ),
        ]);
        assert_eq!(
            router.folder_name(&entry("invoice-receipt.pdf", &[])),
            Some("Receipts")
        );
    }

    #[test]
    fn a_rule_named_like_a_dated_folder_is_refused() {
        let router = RuleRouter::new(&[enabled(
            "bad",
            "2026-08",
            vec![RuleTest::new(Field::Name, "*invoice*")],
        )]);
        assert!(router.is_empty());
        assert!(!router.owns("2026-08"));
    }

    #[test]
    fn decoding_leaves_a_stray_percent_alone() {
        assert_eq!(percent_decode("100%"), "100%");
        assert_eq!(percent_decode("a%zzb"), "a%zzb");
        assert_eq!(percent_decode("a%2"), "a%2");
        assert_eq!(percent_decode("%2Ffacturen%2F"), "/facturen/");
        assert_eq!(percent_decode("no escapes here"), "no escapes here");
        assert_eq!(percent_decode("%C3%A4"), "ä");
    }
}
